pubsub.go

```diff
@@ -359,29 +411,49 @@ func (p *PubsubValueStore) Cancel(name string) (bool, error) {
 	}
 	p.watchLk.Unlock()
 
-	sub, ok := p.subs[name]
+	ti, ok := p.topics[name]
 	if ok {
-		sub.Cancel()
-		delete(p.subs, name)
+		if err := p.closeTopic(name, ti); err != nil {
+			return false, err
+		}
+		<-ti.finished
 	}
 
-	if cancel, ok := p.cancels[name]; ok {
-		cancel()
-		delete(p.cancels, name)
+	return ok, nil
+}
+
+// closeTopic must be called under the PubSubValueStore's mutex
+func (p *PubsubValueStore) closeTopic(key string, ti *topicInfo) error {
+	ti.cancel()
+	ti.sub.Cancel()
+	ti.evts.Cancel()
+	if err := ti.topic.Close(); err != nil {
+		return err
 	}
 
-	return ok, nil
+	delete(p.topics, key)
+	return nil
 }
 
-func (p *PubsubValueStore) handleSubscription(ctx context.Context, sub *pubsub.Subscription, key string, cancel func()) {
-	defer sub.Cancel()
-	defer cancel()
+func (p *PubsubValueStore) handleSubscription(ctx context.Context, ti *topicInfo, key string) {
+	defer func() {
+		ti.finished <- struct{}{}
```

Comment:
The idiom is to close signal channels. That way, waiting twice for some reason (e.g., if closing the topic fails for some reason) doesn't break anything).